pyoptsparse/pyOpt_optimizer.py

```diff
@@ -544,13 +553,15 @@ def _masterFunc2(self, x, evaluate, writeHist=True):
 
                 # Update fail flag
                 masterFail = max(masterFail, fail)
+                self.cache["fail"] = masterFail
 
             # gcon is now in the cache
             returns.append(self.cache["gcon"])
             if self.storeSens:
                 hist["funcsSens"] = self.cache["funcsSens"]
 
-        # Put the fail flag in the history:
+        # Update the fail flag with any cached failure and put the fail flag in the history
+        masterFail = max(self.cache["fail"], masterFail)
```

Comment:
Can we do a single `self.cache["fail"] = masterFail` call here? Instead of having the same line in each of the `if` blocks

Reply:
and do you need to define the dictionary self.cache already with all the keys? can you just not define an empty dicionary?

Reply:
Unfortunately, a single cache set here won't work. In that case, any time it fails, `_masterFunc2` will always return failures since it allows only monotonically increasing fail values (taking the max of the cached and current values every time). I tried a couple other things to avoid duplicate code, but I think that caching the fail value has to happen at the same time the actual user functions are called or else it won't take on the right value.

Reply:
Yeah I see, that makes sense. Some of the duplicated code can probably be pulled out into another function or something, but not needed right now.